File: example/cicada/semigroup.cs
class semigroup_t {
  E : type
  mul : (x: E, y: E) -> E
  associative : (x: E, y: E, z: E) ->
    eqv_t(E, mul(x, mul(y, z)), mul(mul(x, y), z))
}

semigroup : semigroup_t
semigroup = semigroup_t(E, mul, associative)

File: src/main/scala/xieyuheng/structure/semigroup.scala
package xieyuheng.structure

abstract class semigroup_t {
  type elem_t

  def mul(x: elem_t, y: elem_t): elem_t

  def mul_associative(x: elem_t, y: elem_t, z: elem_t) =
    eqv(
      mul(mul(x, y), z),
      mul(x, mul(y, z)))
}
